File: Java/JDI/jdi-uitest-core/src/main/java/com/epam/jdi/uitests/core/reporting/PerformanceStatistic.java
package com.epam.jdi.uitests.core.reporting;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */

import com.epam.commons.map.MapArray;

import java.util.List;

import static com.epam.commons.CalculationUtils.average;

/**
 * Created by Roman_Iovlev on 7/26/2015.
 */
public final class PerformanceStatistic {
    private static MapArray<ActionsType, List<Long>> statistic = new MapArray<>();

    private PerformanceStatistic() {
    }

    public static void addStatistic(long time) {
        addStatistic(ActionsType.JDI_ACTION, time);
    }

    public static void addStatistic(ActionsType actionType, long time) {
        if (statistic.keys().contains(actionType))
            statistic.get(actionType).add(time);
    }

    public static String printStatistic() {
        return "Average Actions Time: " + statistic.toMapArray(value -> Double.toString(average(value)));
    }

}

File: C#.Net/JDI/Core/Reporting/PerformanceStatistic.cs
using System.Collections.Generic;
using System.Linq;
using Epam.JDI.Commons;

namespace Epam.JDI.Core.Reporting
{
    public static class PerformanceStatistic
    {
        private static readonly Dictionary<string, List<double>> Statistic = new Dictionary<string, List<double>>();
        
        public static void AddStatistic(double time)
        {
            AddStatistic("JDI_ACTION", time);
        }

        public static void AddStatistic(string actionType, double time)
        {
            if (Statistic.ContainsKey(actionType))
                Statistic[actionType].Add(time);
        }

        public static string PrintStatistic()
        {
            return "Average Actions Time: " + Statistic.ToDictionary(el => el.Key, el => Average(el.Value).ToString("F")).Print();
        }
        public static double Average(List<double> collection)
        {
            if (collection == null || collection.Count == 0)
                return 0;
            if (collection.Count == 1)
                return collection[0];
            var average = collection[0];
            for (var i = 1; i < collection.Count; i++)
                average = i * (average + collection[i] / i) / (i + 1);
            return average;
        }
    }
}
